engine/lib/filestore.php

```diff
@@ -598,6 +578,209 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = fa
 	return $file_svc->getURL();
 }
 
+/**
+ * Returns a file's URL suitable for embedding in a text editor
+ * We can not use elgg_get_inline_url() for these purposes due to a URL structure bound to user session and file modification time
+ * This function returns a generic (permanent) URL that will then be resolved to an inline URL whenever requested.
+ * The behaviour of the embed-file handler is similar to deprecated mod/file/thumbnail.php
+ *
+ * @param \ElggFile $file File entity (must have guid)
+ * @param string    $size Size
+ * @return string
+ * @since 2.2
+ */
+function elgg_get_embed_url(\ElggFile $file, $size) {
+	return elgg_normalize_url("embed-file/$file->guid/$size");
+}
+
+/**
+ * Page handler for /embed-file/ identifier
+ * The bevaious of this handler is similar to deprecated mod/file/thumbnail.php
+ * 
+ * @param array $segments URL segments
+ * @return void
+ * @access private
+ * @since 2.2
+ */
+function _elgg_filestore_embed_file_handler($segments) {
+
+	$request = _elgg_services()->request;
+	$response = new Symfony\Component\HttpFoundation\Response();
+	$response->prepare($request);
+
+	$guid = array_shift($segments);
+	$size = array_shift($segments);
+
+	$file = get_entity($guid);
+	if (!$file instanceof ElggFile) {
+		if (elgg_entity_exists($guid)) {
+			$response->setStatusCode(403)->setContent('You are not allowed to access this file')->send();
+		} else {
+			$response->setStatusCode(404)->setContent('File does not exist')->send();
+		}
+		exit;
+	}
+
+	$thumbnail = elgg_get_thumbnail($file, $size);
+	if (!$thumbnail->exists()) {
+		$response->setStatusCode(404)->setContent('Thumbnail doest not exist')->send();
+		exit;
+	}
+
+	$if_none_match = $request->headers->get('if_none_match');
+	if (!empty($if_none_match)) {
+		// strip mod_deflate suffixes
+		$request->headers->set('if_none_match', str_replace('-gzip', '', $if_none_match));
+	}
+
+	$filenameonfilestore = $thumbnail->getFilenameOnFilestore();
+	$last_updated = filemtime($filenameonfilestore);
+	$etag = '"' . $last_updated . '"';
+	$response->setPublic()->setEtag($etag);
+	if ($response->isNotModified($request)) {
+		$response->send();
+		exit();
+	}
+
+	header_remove('Cache-Control');
```

Comment:
IIRC I had to remove Cache-Control, Expires, and Pragma; the Symfony components wouldn't replace header, just add them.